File: game/tv_show/in_scoreboard.py
from game_data import GameData
from phone_events import PhoneEvents
from scoreboard.scoreboard_game import ScoreboardGame
from state import State
from tv_show.attract import Attract
from tv_show.going_cave import GoingCave


class InScoreboard(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.scoreboard = ScoreboardGame(context)

    def process_events(self, phone_events: PhoneEvents):
        if phone_events.hungup:
            return Attract

        if self.scoreboard.ended:
            return GoingCave

        self.scoreboard.process_events(phone_events)
        return None

    def render(self, screen):
        self.scoreboard.render(screen)

    def on_exit(self):
        super().on_exit()
        if self.scoreboard is not None and not self.scoreboard.ended:
            self.scoreboard.end()
        self.scoreboard = None

File: game-web/src/tvShow/inScoreboard.ts
/** Port of game/tv_show/in_scoreboard.py */

import { State, type StateClass } from '../state';
import { ScoreboardGame } from '../scoreboard/scoreboardGame';
import { Attract } from './attract';
import { GoingCave } from './goingCave';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class InScoreboard extends State {
  private scoreboard: ScoreboardGame | null;

  constructor(context: GameData) {
    super(context);
    this.scoreboard = new ScoreboardGame(context);
  }

  processEvents(events: PhoneEvents): StateClass | null {
    if (events.hungup) return Attract;
    if (this.scoreboard!.ended) return GoingCave;
    this.scoreboard!.processEvents(events);
    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    this.scoreboard!.render(ctx);
  }

  onExit(): void {
    super.onExit();
    if (this.scoreboard !== null && !this.scoreboard.ended) {
      this.scoreboard.end();
    }
    this.scoreboard = null;
  }
}
